### DISCUSSION TITLE
Message prefix conventions ?

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#71] [Q&A] [answered]
Are there any conventions as to what sort of prefix to use when logging messages to the terminal and history file?  I see *, >, %, *+ and maybe others.  As I expand my use of sdm, build more plugins, and spend more time scanning logs I haven't been consistent in the messages I have created and would probably benefit from conforming to whatever convention is used by the author/expert.

**Comment by gitbls** [comment, marked as answer]
Great question! I tried to establish conventions that make it easy (or at least easier) to scan them for relevant information. In scanning the log messages currently implemented, there are:
```
*+Start of a "section"
* Another Start of a "section" [Best practice]
> Step within a section [Best practice]
>%Warning that something might not be right (Don't use this; will be converting these to "% Warning")
% Another warning [Best practice]
?!Error of some sort (see text below)
? Another error (see text below)
```
There are 250+ messages in sdm (not counting plugins), and there are clearly "improvement opportunities" that I'll get to at some point.

I've noted above those that I consider to be the ones to use, or _best practice_, and will update sdm with these at some point.

For errors, these should be made consistent, but should it be `"? Error"` or `"?!Error"` ? I'm thinking the latter, so it's more obvious when scanning the log. Thoughts?

Hope this helps. As always, appreciate your feedback!

**Reply by dshaw619** [reply]
Thanks for that list.  That helps a lot.  I will modify my scripts as I go to conform (to the best practices).  I appreciate that it might take a LONG time to get all 250+ where you want them.  I'd suggest the list (even in progress) be added as a Doc and then linked from Programming Plugins ..., Burn Scripts, and any other places that discuss such things.

A couple of suggestions:
- If I'm not mistaken, '* ' marks both the Start and End of a section.
- I have a need sometimes to put reminders/instructions in the log re. steps that have to be performed outside sdm (e.g. on another computer).  I've been using a box of '#' around such text, e.g.
```
###################################
# This might be some instructions #
###################################
```
Any better ways?  I've been thinking of maybe appending to a checklist file that could be reviewed after the system is first booted.

re. "? Error" or "?!Error":  pretty much a toss-up, but since you're moving away from the 2 character prefixes elsewhere maybe just '? '.  How about '? ERROR' (and '% WARNING') to help draw attention?

Skimming through my recent logs, here a couple of non-conforming messages that you might consider when you get a chance:
```
% Configure chroot for sdm on your 'x86-64' computer
  * sdm will run more slowly due to qemu emulation
  
>!Ignore 'Running in chroot, ignoring request' message

>%1piboot.conf|auto-1piboot.conf 'boot_behaviour' setting B4 will be honored after First Boot
```

**Comment by gitbls** [comment]
In order (I think!)

- Doc with all the error messages. Could be a good idea, but doesn't seem too manageable, and possibly reduced value due to the number of messages. Upkeep is another issue.
- Yes, '* ' is start and end of "section" to help us all understand what sdm is doing.
- Reminders, etc: There are a lot of ways to implement this. If I needed such a thing I'd probably dump them all to a text file somewhere. 
- Checklist after first boot: As above, write it to a file. You could add a 010-something.sh script to /etc/sdm/0piboot/0*-*.sh that does something during first boot to remind you. For instance, with Lite you can write to the console with `echo -e "\n string" > /dev/console`. Not sure that would work on "with desktop" versions, but if you have a GUI you can set up an autorun that pops the file up in an editor and disables itself from running again. 
- Errors and warnings: I think "? Msg for Error" and "% Msg for Warning" are good, in general. In my mind, the "!" in the 2nd character is supposed to provide a further alert that this msg might not be ignorable.
- Non-conforming messages: I try to have sdm be self-consistent, at least, so will review all the msgs at some point. Thanks for pointing these out.
-

**Reply by dshaw619** [reply]
Sorry if I wasn't clear - I meant just a doc listing the prefix conventions, not every error message (!).  Basically just the list in your first reply edited with any changes in the future or as you transition away from deprecated prefixes.

Thanks again for clarifying all these indicators.

**Comment by gitbls** [comment]
Gotcha. Yes, that's a good idea.

**Reply by gitbls** [reply]
I fixed most, if not all the non-standard messages. Grep is a wonderful thing

**Reply by gitbls** [reply]
Fixed in V7.11
